### DISCUSSION TITLE
How can i remove the cache of all my queries without refetch them again?

### DISCUSSION THREAD

**Discussion by znareak** [TanStack/query#2640] [Q&A]
I want that after the user logout of my application the react query removes its cache, however i was searching in the official docs i found the methods:
![image](https://user-images.githubusercontent.com/22780669/132134998-1ed2883a-7dc1-4c88-a068-7e360be272bb.png)

`queryClient.removeQueries(key);`    <-- Removes the queries ??
`queryClient.invalidateQueries(key);` <-- Invalidate all queries or a specefic query, but refetching it and saves it again in the cache:-/
`queryClient.clear(); `                          <-- Remove the cache?, to me not work this.

![image](https://user-images.githubusercontent.com/22780669/132134946-7a0e1a9a-895b-4f8f-aba0-71649f44e706.png)
The methods not work for me :(

Then, is there some method that react query provides me to remove the cache without refetch them again?
it's a problem the cache because i need that when the user logout, the app remove all user old state and the cache too.

**Comment by TkDodo** [comment]
removeQueries is what you are looking for I believe. Can you show a reproduction of what is not working for you please

**Reply by znareak** [reply]
Thanks for answer me, the error is when a user login , react query does requests and saves them in the cache, but when the user logout, the react query cache still saves the old data and when another user login, the user old data display in the react frontend :(
![react_error](https://user-images.githubusercontent.com/22780669/132137693-162417ab-68ab-4fb8-8e09-8b82bc418316.gif)

I used `queryClient.removeQueries()` but not work for me :(, the old data still display them.
If i `use queryClient.invalidateQueries()` the cache data removes but react query refetching data (without authentication token and that's an error) and saves the data again in the cache.

**Reply by znareak** [reply]
In the example image the user must refresh the page to load the correct user data (card texts) and display in the react UI.

**Reply by TkDodo** [reply]
what I meant was you'd need to provide a reproduction of your issue in the form of, for example, a [codesanbox](https://codesandbox.io/). I can't tell what's going on from looking at a video :)

That being said, `removeQueries` _will_ remove the data from the cache, but it will _not_ inform observers about it. If you still have it rendered on the screen, it will stay there until the component re-renders again.

Here is a very minimal reproduction: https://codesandbox.io/s/xenodochial-leftpad-x1lk3?file=/src/index.js

- click `remove` - you an see in the react-query-devtools panel that the cache is empty
- you can also see that the data is still on the screen
- now increment the counter. because the component re-renders, it will trigger a new fetch, and the data disappears (briefly) because the query will be in `loading` state again

**Reply by znareak** [reply]
I understand, then is there some form to reset the react query internal state?
I don't found some form to reset or that react query delete that state and update the react UI without refresh the page.
I'm trying put that in the logout function to remove all old data user, it's the first time that happens for me i haven't idea of this specific solution.

Sorry for the example, the project has a lot of components and configurations, but in summary i'm using a` useQuery()` hook to fetch user notes and display in the UI. But react query i think that internally has still the old data (because the a SPA no refresh the page) and i am not deleting that old state.

**Reply by znareak** [reply]
I found a similar topic, here: 
https://github.com/tannerlinsley/react-query/discussions/431

**Reply by TkDodo** [reply]
queryClient.resetQueries will reset to the initial state, but it will also refetch in the background

**Reply by znareak** [reply]
Yes i understand, but it's necessary some function resets the state (**delete** all in the cache too) of react query without refetch, because if react-query refetchs the data, then the **fetch results will store in the cache again** but example, if i wish logout an user, then i delete the user token and invalidate the user query, but the react-query will refetch the data (but how don't exist an user token the fetch data is totally invalid) and after the result will store in the cache. 
I'm looking something as this, but it seems that there isn't :(

**Reply by TkDodo** [reply]
if you want to logout a user, `removeQueries` should be enough. Why would you call `invalidateQueries` ? "invalidation" is basically a fancy term for refetch, so with that, you are refetching on purpose...

i feel like we're going in circles here. Please provide a codesandbox of your own to show what you think is not working.

**Comment by mooxl** [comment]
I solved this issue with
```ts
queryClient.clear();
```

**Reply by hamza7681** [reply]
@mooxl It will refetch your data again before redirecting you to unauthorize page

**Reply by cuongnguyenbs123** [reply]
When i logout my app, the dashboard call /me again before it redirect to /login. I think it because my dashboard still mout so the query keep checking the data and refetch

**Comment by busticated** [comment]
this worked for me:

```js
import { onlineManager } from '@tanstack/react-query';

// ...
await queryClient.cancelQueries();
onlineManager.setOnline(false);
queryClient.clear();
// perform logout, redirect, etc
```

`onlineManager.setOnline(false);` has the effect of preventing all fetches / network calls which is exactly what i want. unfortunately, it has the side-effect of triggering any subscribers added via `onlineManager.subscribe()`. for my app in its current state, this trade-off is fine but it's definitely not ideal.

i'm aware of the query `enabled` field and related. using that would require a fair amount of plumbing to ensure _all_ queries are coordinated. afaict `queryClient.removeQueries()` doesn't restrain currently mounted components w/ `useQuery()` calls.

it would be great if there was a simple, global way to achieve what `onlineManager.setOnline(false);` does without also implying a lack of network 🙏 👍
